DOC Eager loading (#277)
silverstripe/developer-docs: en/02_Developer_Guides/00_Model/02_Relations.md

```sql
# Retrieve all the teams
# Note this is not the exact SQL that would generated it is just for demonstration
SELECT * FROM Team;

# Retrieve the players for all the teams in 20 separate queries:
SELECT * FROM Player WHERE TeamID = 1
SELECT * FROM Player WHERE TeamID = 2
SELECT * FROM Player WHERE TeamID = 3
SELECT * FROM Player WHERE TeamID = ...
```

The N + 1 query problem can be alleviated using eager loading which in this example will reduce this down to just two queries. We do this by passing the relationships that should be eagerly loaded to the `eagerLoad()` method on [DataList](api:SilverStripe\ORM\DataList):

```php
$teams = Team::get()->eagerLoad('Players')
```

With eager loading now only two queries will be executed:

```sql
# Retrieve all the teams
SELECT * FROM Team

# Retrieve all the players for the teams in a single query:
SELECT * FROM Player WHERE TeamID IN (1, 2, 3, ...)
```

Eager load multiple relationships by passing additional arguments to the ->eagerLoad() method:

```php
$teams = Team::get()->eagerLoad('Players', 'Fans');

> ```

Reviewer: Could we provide a little bit more details about `multiple relationships` and  `nested relationships`. 
Probably, some code example to show difference. 
And also could we clarify how many `additional arguments` might be passed, for example, like below _"up to **three** levels deep"_, so it means you can request something `Team::get()->eagerLoad('Players', 'Fans', 'Games')` or `Team::get()->eagerLoad('Players.Games', 'Fans')`

Author: Updated

> Eager load multiple relationships by passing additional arguments separated by a comma to the `eagerLoad()` method. In this example `Players` and `Fans` are `has_many` relationships on the `Team` model:

Reviewer: ```suggestion
Suppose we have the following architecture and relationships between classes:
```php
class Team extends DataObject
{
    private static $has_many = [
        'Players' => Player::class,
        'Fans' => Fan::class,
    ];
}

class Player extends DataObject
{
    private static $has_one = [
        'Team' => Team::class,
    ];
    
    private static $many_many = [
        'Games' => Game::class,
    ];
}

class Game extends DataObject
{
    private static $has_many = [
         'Players' => Player::class,
        'Officials' => Official::class,
        'Sponsors' => Sponsor::class,
    ];
}
In this example, to get all Players and Fans, you can use eager loading multiple relationships by passing additional arguments separated by a comma to the `eagerLoad()` method:
```

Author: Updated

> Eager load nested relationships up to **three** levels deep using the dot syntax. Three levels deep means there will be 2 dots in the relationship. In this example `Games` is a `has_many` relationship on the `Player` model, and `Officials` is a `has_many` relationship on the `Game` model:
> 
> ```php
> $teams = Team::get()->eagerLoad('Players.Games.Officials')
> 

Reviewer: ```suggestion
Perhaps, you may need to get all the Officials that are related to this Game. In this example, you can use another feature provided by the `eagerLoad` method. Eager load nested relationships up to **three** levels deep using the dot syntax:

```php
$teams = Team::get()->eagerLoad('Players.Games.Officials');

You can even get the results for multiple nested relations by passing them with a comma:

```php
$teams = Team::get()->eagerLoad(
          'Players.Games.Officials', 
          'Players.Games.Sponsors', 
);
```

> foreach ($teams as $team) {
>     foreach ($team->Players() as $player) {
>         foreach ($player->Games() as $game) {
>             foreach ($game->Officials() as $official) {
>                 // Everything will have been eager loaded at this point
>                 echo $official->FirstName;
>             }
>         }
>     }
> }
> ```

Reviewer: ```suggestion
After receiving the result, you can work with it like with a regular associative array:
```php
foreach ($teams as $team) {
    foreach ($team->Players() as $player) {
        foreach ($player->Games() as $game) {
            foreach ($game->Officials() as $official) {
                // Everything will have been eager loaded at this point
                echo $official->FirstName;
            }
        }
    }
}
```
```
